File: nlp-task-dispatcher/src/main/java/com/example/nlptaskdispatcher/redis/RedisClient.java
package com.example.nlptaskdispatcher.redis;

import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;

import java.util.Map;

public class RedisClient {

    private static final JedisPool jedisPool = new JedisPool("redis", 6379);

    private static final String NLP_TASK_QUEUE = "nlp_task_queue";

    public static Jedis getResource() {
        return jedisPool.getResource();
    }

    public static void pushTask(String taskJson) {
        try (Jedis jedis = getResource()) {
            jedis.lpush(NLP_TASK_QUEUE, taskJson);
        }
    }

    public static String popTaskBlocking(int timeoutSeconds) {
        try (Jedis jedis = getResource()) {
            var result = jedis.brpop(timeoutSeconds, NLP_TASK_QUEUE);
            if (result != null && result.size() > 1) {
                return result.get(1);
            } else {
                return null;
            }
        }
    }

    public static void setTaskOverallStatus(String taskId, String status) {
        try (Jedis jedis = jedisPool.getResource()) {
            jedis.setex("task: " + taskId, 3600, status);
        }
    }

    public static String getTaskOverallStatus(String taskId) {
        try (Jedis jedis = jedisPool.getResource()) {
            return jedis.get("task: " + taskId);
        }
    }

    public static void setTaskStatus(String taskId, String articleId, Map<String, String> fields, int expireSeconds) {
        String key = String.format("task:%s:%s", taskId, articleId);
        try (Jedis jedis = getResource()) {
            jedis.hset(key, fields);
            jedis.expire(key, expireSeconds);
        }
    }

    public static Map<String, String> getTaskStatus(String taskId, String articleId) {
        String key = String.format("task:%s:%s", taskId, articleId);
        try (Jedis jedis = getResource()) {
            return jedis.hgetAll(key);
        }
    }

    public static void deleteTaskStatus(String taskId, String articleId) {
        String key = String.format("task:%s:%s", taskId, articleId);
        try (Jedis jedis = getResource()) {
            jedis.del(key);
        }
    }
}

File: python-nlp/app/redis_client.py
import redis
import json

# r = redis.Redis(host='localhost', port=6379, db=0, decode_responses=True)
r = redis.Redis(host='redis', port=6379, db=0, decode_responses=True)

QUEUE_NAME = "nlp_task_queue"

def set_task_status(task_id: str, article_id: str, status: str = None, extra: dict = None):
    key = f"task:{task_id}:{article_id}"
    value = {}

    if status is not None:
        value["status"] = status

    if extra:
        for k, v in extra.items():
            if isinstance(v, (dict, list)):
                value[k] = json.dumps(v, ensure_ascii=False)
            else:
                value[k] = str(v)

    if value:
        r.hset(key, mapping=value)
        r.expire(key, 60)

def get_task_status(task_id: str, article_id: str) -> dict:
    key = f"task:{task_id}:{article_id}"
    return r.hgetall(key)

def push_task(task: dict):
    task_json = json.dumps(task, ensure_ascii=False)
    r.lpush(QUEUE_NAME, task_json)

def pop_task_blocking(timeout: int = 5) -> dict | None:
    result = r.brpop(QUEUE_NAME, timeout)
    if result is None:
        return None
    _, task_json = result
    return json.loads(task_json)
